### ISSUE TITLE
`[.]Type` completion for `(any P).` shows it will produce an existential metatype

### ISSUE THREAD

**Issue by AnthonyLatsis** [apple/swift#65843]
Test case:
```swift
// RUN: %empty-directory(%t)
// RUN: %target-swift-ide-test -batch-code-completion -source-filename %s -filecheck %raw-FileCheck -completion-output-dir %t
protocol P {}

(any P).#^COMPLETE?check=META^# 
// META: Keyword/CurrNominal:  Type[#(any P).Type#]; name=Type
```

This test will fail, because the current completion is `Type[#any P.Type#]; name=Type`. Notice how it thinks the completion will produce an existential metatype `any P.Type`, whereas the actual result is the singleton metatype `(any P).Type`.

**Environment**
- Swift 5.9-dev (c1d5118c21da49a9c2194734cbf9b6970bc63056)

**Comment by AnthonyLatsis** [comment]
@Rajveer100 This one is somewhat less involved if you’re interested.

**Comment by AnthonyLatsis** [comment]
Yeah, I think that's a good idea.

**Comment by Rajveer100** [comment]
@AnthonyLatsis What could be the cause for this?

**Comment by AnthonyLatsis** [comment]
Stale build artifacts, for example. Rule no. 1 (assuming your checkout revisions are in sync): try a clean build.

**Comment by AnthonyLatsis** [comment]
I can't tell much about what went wrong without more context. Can you share the steps you took originally and after my suggestion?

**Comment by Rajveer100** [comment]
I mean the only thing I did is change my target scheme to swift-ide-test, as you said to test for code completion.

**Comment by AnthonyLatsis** [comment]
Ah, so you're getting this error in Xcode, not as a result of invoking `build-script`. Does it build fine if you switch back to `swift-frontend`? Could you have selected a generated `swift-ide-test` target/scheme instead of manually creating your own, as you did for `swift-frontend` per the guide?

**Comment by Rajveer100** [comment]
Oh yeah, I used the existing one!

**Comment by Rajveer100** [comment]
So I need to create a new scheme with the target as swift-use-test? Any other configurations that's needed like it was in swift-frontend?

**Comment by AnthonyLatsis** [comment]
You need to follow the exact same steps, but for the `swift-ide-test` executable target.

**Comment by Rajveer100** [comment]
Now it builds successfully and does run with the following output:

```
action required
OVERVIEW: Swift IDE Test

USAGE: swift-ide-test [options] [input files...]

...
```

**Comment by Rajveer100** [comment]
Where do I use the [options] so I can accordingly check for the specific token?

**Comment by Rajveer100** [comment]
@AnthonyLatsis I added the following as `args...` in the scheme for `swift-ide-test`,  what would be the file that's required as per the log message:

`swift-ide-test --code-completion --code-completion-token="#^COMPLETE?check=META^#"`

log: `source file required`

**Comment by AnthonyLatsis** [comment]
Good job puzzling out how to output completions for a single token! This tool requires a path to an input source file to read from: `-source-filename <path to file>`. Also, `-code-completion-token` expects a token name/id like `COMPLETE`, not the token itself.

**Comment by Rajveer100** [comment]
Got this:

```
found code completion token COMPLETE at offset 951
Begin completions, 3 items
Keyword[self]/CurrNominal:          self[#(any P).Type#]; name=self
Keyword/CurrNominal:                Protocol[#(any P).Type#]; name=Protocol
Keyword/CurrNominal:                Type[#any P.Type#]; name=Type
End completions
LookedupTypeNames: ['swift_ide_test.P']
```

**Comment by Rajveer100** [comment]
I am aware of `existential meta types`, but could you differentiate it from `singleton meta types` and also brief me out the log message which I obtained above. Regarding the files, I suppose `CompletionLookup::getTypeCompletions` seems to be the place to debug.

**Comment by AnthonyLatsis** [comment]
The *singleton metatype* is simply the metatype of a given type `T` written as `T.Type` (e.g. `Int.Type`, `(any P).Type`). The only possible instance of a Swift metatype is the type object `T.self`, hence *singleton*. An existential metatype is not really a metatype, but is called as such because it can be assigned type objects. If `any P` is a box that wraps a value of dynamic type that conforms to `P`, then `any P.Type`, aka an *existential metatype*, is a box that wraps a value of dynamic metatype whose instance type conforms to `P`. So if `Int` conforms to `P`, then `any P` can wrap `0`, and `any P.Type` can wrap `Int.self`. 

___

The places to debug should be accurate.

**Comment by Rajveer100** [comment]
Yeah that gave me a nice idea! Let me try to debug!

**Comment by Rajveer100** [comment]
@AnthonyLatsis
Under `CompletionLookup::getPostfixKeywordCompletions`, we convert the `ParsedExpr` to a type of `AnyMetatypeType`,
which is where the check for `isExistentialType()` succeeds and the following snippet probably causes it:

```
// Just below "Protocol" keyword addition
addKeyword("Type", ExistentialMetatypeType::get(instanceTy),
                   SemanticContextKind::CurrentNominal);
```

Is my intuition right? If yes, any hints?

**Comment by Rajveer100** [comment]
Do we want an extra brace `{}` surrounded for the meta type before looking up for the `BaseType`?

**Comment by Rajveer100** [comment]
@AnthonyLatsis 
Am I on the right track here?

**Comment by AnthonyLatsis** [comment]
> which is where the check for `isExistentialType()` succeeds and the following snippet probably causes it

Yes, this very much looks like the cause. You have to understand that `any P.Type` and `(any P).Type` are totally different types *semantically*. It is not that one type simply has an additional layer of parentheses as in `Int` vs. `(Int)`:

* `base.Type` → `MetatypeType(base)` — a regular, singleton, metatype type.
*  `any base.Type` → `ExistentialMetatypeType(base)` — an existential metatype type. 

Hence `(any P).Type` → `MetatypeType(any P)` and `any P.Type` → `ExistentialMetatypeType(P)`.

**Comment by Rajveer100** [comment]
In this function:

```C++
inline bool CanType::isExistentialTypeImpl(CanType type) {
  return isa<ProtocolType>(type) ||
         isa<ProtocolCompositionType>(type) ||
         isa<ExistentialType>(type) ||
         isa<ParameterizedProtocolType>(type);
}
```

Why are Protocols considered as `isExistentialTypeImpl`?

**Comment by AnthonyLatsis** [comment]
For no good reason. That this has not been known to cause problems is a slightly worrying fluke. Before a separate representation for existential types was adopted in the model, this function worked with what is now known as a constraint type — protocol and protocol composition types. Then (apparently during the transition) someone added the `isa<ExistentialType>(type)` term . It might have been an intentional technical debt.

**Comment by Rajveer100** [comment]
At the moment I feel there could only be the following possibilities:

- Either the problem lies in the check for existential type.
- Or when get the expression by using `getAs`.

**Comment by AnthonyLatsis** [comment]
The current logic assumes that a postfix `Type` completion will necessarily produce an existential metatype if the instance type of the parsed type expression `isAnyExistentialType()`. This is true for `any P` and false for `(any P)`: the parentheses do not affect the canonical type of the expression, but they do change the outcome of appending `Type`.

**Comment by Rajveer100** [comment]
If the canonical type remains the same shouldn't this check succeed:

```C++
!instanceTy->getCanonicalType()->is<MetatypeType>()
```

**Comment by AnthonyLatsis** [comment]
I’m not catching on to the causal relationship, but yes, the instance type is not a metatype in either of the above cases.

**Comment by AnthonyLatsis** [comment]
Let me give the explanation another shot. If you feel unconfident around `any`/metatypes, you may want to put a pause on this and do some research on them.

*  `(any P).Type` (the singleton metatype of existential type `any P`) ≠ `any P.Type` = `P.Type` (an existential metatype constrained to protocol `P`).
* For `P.`, the correct `Type` completion is `Type[#any P.Type#]` because `any P.Type` is the modern way of spelling `P.Type` = `P.` + `Type`.
* Ditto `any P.`. In this case, the result is exactly `any P.Type`.
* For `(any P).`, the correct `Type` completion is `Type[#(any P).Type#]` because `(any P).` + `Type` = `(any P).Type`.

**Comment by Jamezzzb** [comment]
Also from the [SE-0335](https://github.com/apple/swift-evolution/blob/main/proposals/0335-existential-any.md) .md: Spell the existential metatype as `any P.Type`, and the protocol metatype as `(any P).Type`.

Reiterating part of the above: `any P` and `(any P)` are the same but `any P.Type` and `(any P).Type` are different. Which is what the current completion logic seems to be missing out on.
